File: USR-CoreFab/src/isolation.rs
//! Constitutional Isolation — Capability-based sandboxing of components.
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use crate::{Result, ConstitutionalError, timestamp};

/// A capability grant
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Capability {
    pub resource: String,
    pub permissions: Vec<String>,
    pub expires: String,
    pub granted_to: String,
    pub granted_at: String,
}

/// Resource types that can be governed
pub const GOVERNED_RESOURCES: &[&str] = &[
    "execution",
    "attestation",
    "orchestration",
    "identity",
    "constraints",
    "storage",
    "network",
];

/// Capability-based isolation engine
pub struct IsolationEngine {
    granted: HashMap<String, Vec<Capability>>,
}

impl IsolationEngine {
    pub fn new() -> Self {
        Self {
            granted: HashMap::new(),
        }
    }

    /// Request a capability from the runtime
    pub fn request_capability(
        &mut self,
        component_id: &str,
        resource: &str,
        permissions: &[String],
    ) -> Result<Capability> {
        // Validate the resource is governable
        if !GOVERNED_RESOURCES.contains(&resource) {
            return Err(ConstitutionalError::CapabilityNotGranted(
                format!("Resource '{}' is not governed by the runtime", resource)
            ));
        }

        // For v0.1.0, all requests are granted for constitutional components
        // In a full implementation, this would check component identity and layer
        let capability = Capability {
            resource: resource.to_string(),
            permissions: permissions.to_vec(),
            expires: "permanent".to_string(),
            granted_to: component_id.to_string(),
            granted_at: timestamp(),
        };

        self.granted
            .entry(component_id.to_string())
            .or_insert_with(Vec::new)
            .push(capability.clone());

        Ok(capability)
    }

    /// Check if a capability is currently granted
    pub fn check_capability(
        &self,
        component_id: &str,
        resource: &str,
        permission: &str,
    ) -> Result<bool> {
        if let Some(capabilities) = self.granted.get(component_id) {
            for cap in capabilities {
                if cap.resource == resource && cap.permissions.contains(&permission.to_string()) {
                    // Check expiration
                    if cap.expires == "permanent" || cap.expires > timestamp() {
                        return Ok(true);
                    }
                }
            }
        }
        Ok(false)
    }

    /// Revoke a capability
    pub fn revoke_capability(
        &mut self,
        component_id: &str,
        resource: &str,
    ) -> Result<bool> {
        if let Some(capabilities) = self.granted.get_mut(component_id) {
            let before = capabilities.len();
            capabilities.retain(|c| c.resource != resource);
            return Ok(capabilities.len() < before);
        }
        Ok(false)
    }

    /// Get all capabilities for a component
    pub fn get_capabilities(&self, component_id: &str) -> Vec<&Capability> {
        self.granted
            .get(component_id)
            .map(|caps| caps.iter().collect())
            .unwrap_or_default()
    }

    /// Verify all capabilities are valid
    pub fn verify_all(&self) -> Result<bool> {
        let now = timestamp();
        for (component_id, capabilities) in &self.granted {
            for cap in capabilities {
                if cap.expires != "permanent" && cap.expires < now {
                    return Err(ConstitutionalError::CapabilityNotGranted(
                        format!("Expired capability for {} on {}", component_id, cap.resource)
                    ));
                }
            }
        }
        Ok(true)
    }
}

File: usr/engine/isolation.js
// USR/CoreFab — Isolation Contract (C4)
// Constitutional Isolation Engine — capability-based sandboxing.

const GOVERNED_RESOURCES = [
  'execution', 'attestation', 'orchestration',
  'identity', 'constraints', 'storage', 'network',
];

class IsolationEngine {
  constructor() {
    this.granted = new Map();
  }

  async requestCapability(componentId, resource, permissions) {
    if (!GOVERNED_RESOURCES.includes(resource)) {
      throw new Error(`CapabilityNotGranted: Resource '${resource}' is not governed`);
    }

    const capability = {
      resource,
      permissions: Array.isArray(permissions) ? permissions : [permissions],
      expires: 'permanent',
      grantedTo: componentId,
      grantedAt: timestamp(),
    };

    if (!this.granted.has(componentId)) this.granted.set(componentId, []);
    this.granted.get(componentId).push(capability);

    window.dispatchEvent(new CustomEvent('usr:isolation:granted', { detail: { componentId, capability } }));
    return capability;
  }

  checkCapability(componentId, resource, permission) {
    const caps = this.granted.get(componentId) || [];
    for (const cap of caps) {
      if (cap.resource === resource && cap.permissions.includes(permission)) {
        if (cap.expires === 'permanent' || cap.expires > timestamp()) return true;
      }
    }
    return false;
  }

  revokeCapability(componentId, resource) {
    const caps = this.granted.get(componentId) || [];
    const before = caps.length;
    const filtered = caps.filter(c => c.resource !== resource);
    this.granted.set(componentId, filtered);
    const revoked = filtered.length < before;
    if (revoked) window.dispatchEvent(new CustomEvent('usr:isolation:revoked', { detail: { componentId, resource } }));
    return revoked;
  }

  getCapabilities(componentId) { return this.granted.get(componentId) || []; }

  async verifyAll() {
    const now = timestamp();
    for (const [id, caps] of this.granted) {
      for (const cap of caps) {
        if (cap.expires !== 'permanent' && cap.expires < now) {
          throw new Error(`CapabilityNotGranted: Expired capability for ${id} on ${cap.resource}`);
        }
      }
    }
    return true;
  }
}

window.IsolationEngine = IsolationEngine;
window.GOVERNED_RESOURCES = GOVERNED_RESOURCES;
